src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceReader.scala

```diff
@@ -0,0 +1,134 @@
+
+  override def pruneColumns(schema: StructType): Unit = requiredSchema = Some(schema)
+
+  private case class MongoDataReaderFactory(connector: MongoConnector, partition: MongoPartition, filters: Array[Filter],
+                                            requiredSchema: Option[StructType]) extends DataReaderFactory[Row] with LoggingTrait {
+    override def createDataReader(): DataReader[Row] = {
+      MongoDataReader(connector, partition, createPipeline())
+    }
+
+    private def createPipeline(): Seq[Bson] = {
+      if (readConfig.pipelineIncludeFiltersAndProjections) {
+        val pipelineFilters = if (readConfig.pipelineIncludeNullFilters) {
+          schema.fields.filter(!_.nullable).map(_.name).map(IsNotNull) ++ filters
+        } else {
+          filters
+        }
+        val requiredColumns = requiredSchema.map(_.fieldNames).getOrElse(Array.empty[String])
+        if (requiredColumns.nonEmpty || filters.nonEmpty) {
+          logInfo(s"requiredColumns: ${requiredColumns.mkString(", ")}, filters: ${pipelineFilters.mkString(", ")}")
+        }
+
+        pipeline ++ MongoRelationHelper.createPipeline(requiredColumns, pipelineFilters)
+      } else {
+        logInfo(s"${ReadConfig.pipelineIncludeFiltersAndProjectionsProperty} set to false. Filters and Projections have not been pushed to MongoDB")
+        pipeline
+      }
+    }
+  }
+
+  private case class MongoDataReader(connector: MongoConnector, partition: MongoPartition, pipeline: Seq[Bson]) extends DataReader[Row] {
+    var _mongoClient: Option[MongoClient] = None
+    var _cursor: Option[MongoCursor[BsonDocument]] = None
+
+    def mongoClient: MongoClient = {
+      if (_mongoClient.isEmpty) {
+        _mongoClient = Some(connector.acquireClient())
+      }
+      _mongoClient.get
+    }
+
+    def cursor: MongoCursor[BsonDocument] = {
+      if (_cursor.isEmpty) {
+
+        val partitionPipeline: Seq[Bson] = if (partition.queryBounds.isEmpty) {
+          pipeline
+        } else {
+          new BsonDocument("$match", partition.queryBounds) +: pipeline
+        }
+
+        val aggregateIterable: AggregateIterable[BsonDocument] = mongoClient.getDatabase(readConfig.databaseName)
+          .getCollection[BsonDocument](readConfig.collectionName, classOf[BsonDocument])
+          .withReadConcern(readConfig.readConcern)
+          .withReadPreference(readConfig.readPreference)
+          .aggregate(partitionPipeline.asJava)
+          .allowDiskUse(readConfig.aggregationConfig.allowDiskUse)
+
+        readConfig.aggregationConfig.hint.map(aggregateIterable.hint)
+        readConfig.aggregationConfig.collation.map(aggregateIterable.collation)
+        _cursor = Some(aggregateIterable.iterator)
+      }
+      _cursor.get
+    }
+
+    override def next(): Boolean = cursor.hasNext
+
+    override def get(): Row = documentToRow(cursor.next(), readSchema())
+
+    override def close(): Unit = {
+      _cursor.foreach(_.close())
```

Comment:
foreach seems an odd choice for the Option here - but as foreach returns Unit its considered better than using map which is expected to return a value.